docs/containerization-design-doc.md

```diff
@@ -0,0 +1,384 @@
+     Response:
+     Creates ramp up records
+     Status: 201 created  
+```
+
+At dispatch time, a graph walk will be performed to find out all the job types that
+the flow intends to execute. Their "default" version will be picked from the database
+table. Users can override the default version through runtime properties. The version
+maps to the specific Image. Details are described in the [Dispatch Logic Section](#dispatch-logic).
+
+### Dispatch Logic 
+ 
+1. Whenever a flow is ready to run (By schedule, by data triggers or manually through UI/API call),
+the AZ Web Server will mark the flow with the state: ```READY``` and insert the flow in the queue
+(**execution_flows** table).
+ 
+2. The **QueueProcessor** class will pick up executions based on priority and submit_time and set the state as
+```DISPATCHING``` for the picked flows. A rate-limiter is introduced here so the Kubernetes namespace does not
+get overwhelmed with the rate of creating containers. Finally, each of the picked flows are then submitted for
+**dispatch**. The dispatch logic:
+
+    * Will require a pre-configured Kubeconfig file to organize the information about clusters, users, namespaces
+    and authentication mechanism. Kubeconfig file for Azkaban-web-server will contain the path for the certificate
+    which will be used to authenticate with Kubernetes. Azkaban web server needs
+    permission to create pods in a namespace dedicated for containerization on each Kubernetes cluster.
+    
+    * A graph walk on the flow is performed to determine which jobtypes are needed for the execution environment.
+    **image_ramp_up** and the **image_versions** tables are used to identify which image versions should be
+    used for Azkaban, platform and the corresponding job types.
+    
+    * The needed versions Config are put together in alphabetical order in a JSON file and it constitutes the
+    **Version-Set** that is used to create the execution environment. The Version Set constitutes all details
+    required to reconstruct the execution environment if a repeat execution is required. If the version set
+    is not already existing in tables, is added to **version_set** table.
+    
+      * The version_set is a very useful concept. With the proposed changes, Azkaban will become a very dynamic
+      environment where the list of binaries used to launch a flow will become very dynamic (Some combination of
+      Platform binaries, Azkaban binaries or job type binaries are always ramping up). In such an environment,
+      in order to debug flows it may become important to reproduce the same set of binaries, which were used to
+      launch the flow. The version_set number provides a single version string that represents the list of
+      binaries with their corresponding version numbers to launch the flow. Hence, reproducible!
+    
+    * Finally, a YAML file is constructed on the fly for all parameters necessary to create the execution
```

Comment:
There will be some change required for this paragraph based on pod spec builder changes that Aditya has introduced. Just FYI. But we can update it later.